Memories/Memories/VCs/AddViewController.swift

```diff
@@ -0,0 +1,159 @@
+    @IBOutlet weak var titleField: UITextField!
+    @IBOutlet weak var imageButton: UIButton!
+    @IBOutlet weak var filterButton: UIButton!
+    
+    @IBOutlet weak var audioButton: UIButton!
+    @IBOutlet weak var videoButton: UIButton!
+    @IBOutlet weak var audioView: UIView!
+    @IBOutlet weak var videoView: UIView!
+    
+    var titleText: String?
+    var image: Data?
+    var originalImage: Data?
+    var audio: String? { UserDefaults.standard.string(forKey: "CurrentAudio") }
+    var video: String? { UserDefaults.standard.string(forKey: "CurrentVideo") }
+    
+    var picker = UIImagePickerController()
+    let locationManager = CLLocationManager()
+    var coordinates: CLLocationCoordinate2D? { didSet { finishSaving() }}
+    private let context = CIContext(options: nil)
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        picker.delegate = self
+        UserDefaults.standard.removeObject(forKey: "CurrentAudio")
+        UserDefaults.standard.removeObject(forKey: "CurrentVideo")
+    }
+    
+    @IBAction func textFieldReturn(_ sender: UITextField) {
+        sender.resignFirstResponder()
+    }
+    
+    @IBAction func addImage(_ sender: Any) {
+        picker.allowsEditing = true
+        let actionSheet = UIAlertController(title: "Photo Source", message: "Choose a source", preferredStyle: .actionSheet)
+        
+        actionSheet.addAction(UIAlertAction(title: "Camera", style: .default , handler: { (sction: UIAlertAction) in
+            if UIImagePickerController.isSourceTypeAvailable(.camera) {
+                self.picker.sourceType = .camera
+                self.present(self.picker, animated: true, completion: nil) }
+            else { print("Camera not available") } }))
+        actionSheet.addAction(UIAlertAction(title: "Photo Libary", style: .default , handler: { (sction: UIAlertAction) in
+            self.picker.sourceType = .photoLibrary
+            self.present(self.picker, animated: true, completion: nil) }))
+        actionSheet.addAction(UIAlertAction(title: "Cancel", style: .cancel, handler: nil))
+        self.present(actionSheet, animated: true)
+    }
+    
+    @IBAction func filterImage(_ sender: Any) {
+        filterButton.isSelected.toggle()
+        if filterButton.isSelected {
+        guard let image = originalImage else { return }
+        var ciImage = CIImage(data: image)
+        let filter = CIFilter.colorMonochrome()
+        filter.inputImage = ciImage
+        filter.color = CIColor(color: .black)
+        ciImage = filter.outputImage
+        guard let outputCI = ciImage else { return }
+        guard let outputCG = context.createCGImage(outputCI, from: CGRect(origin: CGPoint.zero, size: imageView.image!.size)) else { return }
+        let uiImage = UIImage(cgImage: outputCG)
+        imageView.image = uiImage
+            self.image = uiImage.pngData()
+        }
+        else {
+            imageView.image = UIImage(data: originalImage!)
+        }
+    }
```

Comment:
1. Please make IBOutlets have more meaningful names so you can create helper functions to support them (filterButtonPressed)
2. Add helper functions like `func filterImage(_ image: UIImage) -> UIImage?`
3. Store actual images (not data), so you don't reach into the UI to get data that should be part of a model, or temporary model attribute. (Don't use `imageView.image!.size`)
Here's how I might write it
4. Don't force unwrap originalImage, it's risky and prone to crash at runtime: `imageView.image = UIImage(data: originalImage!)`
```swift
@IBAction func filterImage(_ sender: Any) {
    filterButton.isSelected.toggle()
    if filterButton.isSelected {
        
        // names are confusing because originalImage is really originalImageData
        guard let imageData = originalImage else { return }
        guard let image = filterImageData(imageData) else { return }
        
        // separate UI logic (filtering on main thread will slow UI)
        imageView.image = image
        self.image = image.pngData()
    }
    else {
        imageView.image = UIImage(data: originalImage!) // FIXME: don't force unwrap
    }
}

// FIXME: change originalImage so it's an image, not data, we lose info like image.size
// without storing UIImage types
private func filterImageData(_ imageData: Data) -> UIImage? {
    let ciImage = CIImage(data: imageData)
    let filter = CIFilter.colorMonochrome()
    filter.inputImage = ciImage
    filter.color = CIColor(color: .gray) // gray gives some nice values than .black
//        ciImage = filter.outputImage  // Don't reuse the ciImage here on output
    guard let outputCI = filter.outputImage else { return nil } // use it here
    
    // Don't use the imageView.image! to get the size, you should store the actual image rather than data,
    // so that you can pass in an image, not imageData (Data)
    guard let outputCG = context.createCGImage(outputCI, from: CGRect(origin: CGPoint.zero, size: imageView.image!.size)) else { return nil }
    
    let uiImage = UIImage(cgImage: outputCG)
    return uiImage
}
```